### PR TITLE
Make has_unquote/1 quote-aware

### PR DESCRIPTION
**Pull request by sabiwara**
Only consider top-level unquotes, as discussed [here](https://github.com/elixir-lang/elixir/pull/12832#discussion_r1284094065).

### CODE REVIEW

File: lib/elixir/test/elixir/kernel/has_unquote_test.exs [hunk 1]
```diff
@@ -0,0 +1,122 @@
+Code.require_file("../test_helper.exs", __DIR__)
+
+defmodule Kernel.HasUnquoteTest do
```
Comment (sabiwara) [inline lib/elixir/test/elixir/kernel/has_unquote_test.exs:3]: Feel free to suggest a better place for these tests.

I know `has_unquotes/1` is purely internal, but given its complexity and edge cases, unit-testing feels highly valuable to validate and reason about its expected behavior.

File: lib/elixir/src/elixir_quote.erl [hunk 2]
```diff
@@ -30,17 +30,35 @@ fun_to_quoted(Function) ->
...
+
+has_unquotes({quote, _, [Child]}, QuoteLevel) ->
+  has_unquotes(Child, QuoteLevel + 1);
+has_unquotes({quote, _, [QuoteOpts, Child]}, QuoteLevel) ->
+  case disables_unquote(QuoteOpts) of
+    true -> false;
+    _ -> has_unquotes(Child, QuoteLevel + 1)
+  end;
+has_unquotes({Unquote, _, [Child]}, QuoteLevel)
+  when (Unquote == unquote) orelse (Unquote == unquote_splicing) ->
```
Comment (josevalim) [inline lib/elixir/src/elixir_quote.erl:43]:
```suggestion
  when Unquote == unquote; Unquote == unquote_splicing ->
```

File: lib/elixir/test/elixir/kernel/has_unquote_test.exs (hunk 1 above)
Comment (josevalim) [reply lib/elixir/test/elixir/kernel/has_unquote_test.exs:3]: I would move to kernel/quote_test.exs :)

**Review by josevalim — APPROVED** [review]

File: lib/elixir/src/elixir_quote.erl (hunk 2 above)
Comment (sabiwara) [reply lib/elixir/src/elixir_quote.erl]: Thank you, much better.
The `,` vs `;` in Erlang guards always trips me up  😅

File: lib/elixir/test/elixir/kernel/has_unquote_test.exs (hunk 1 above)
Comment (sabiwara) [reply lib/elixir/test/elixir/kernel/has_unquote_test.exs]: Perfect! Done.

**Comment by sabiwara** [comment]
Rebased and applied suggestions.

**Review by josevalim — APPROVED** [review]

**Pull request merged**